lib/api-tests/tests/high_level_api.rs

```diff
@@ -0,0 +1,177 @@
+static TEST_WAT: &str = r#"
+(module
+  (import "env" "outside-global" (global $outside-global (mut i32)))
+  (import "env" "update-func" (func $update-func (param i32) (result i32)))
+  (table $test-table (export "test-table") 2 anyfunc)
+  (global $test-global (export "test-global") (mut i32) (i32.const 3))
+  (elem (;0;) (i32.const 0) $ret_2)
+  (func $ret_2 (export "ret_2") (result i32)
+    i32.const 2)
+  (func $ret_4 (export "ret_4") (result i32)
+    i32.const 4)
+  (func $set_test_global (export "set_test_global") (param i32)
+    (global.set $test-global
+                (local.get 0)))
+  (func $update-outside-global (export "update_outside_global")
+    (global.set $outside-global
+                (call $update-func (global.get $outside-global))))
+)
+"#;
+
+fn append_custom_section(
+    wasm: &[u8],
+    custom_section_name: &str,
+    custom_section_contents: &[u8],
+) -> Vec<u8> {
+    let mut out = Vec::with_capacity(
+        // 1 for custom section id, 5 for max length of custom section, 5 for max length of name
+        wasm.len() + custom_section_name.len() + custom_section_contents.len() + 1 + 5 + 5,
+    );
+
+    out.extend(wasm);
+    out.push(0);
+
+    let name_length = custom_section_name.as_bytes().len() as u32;
+    let mut name_length_as_leb128 = vec![];
+    write_u32_leb128(&mut name_length_as_leb128, name_length);
+
+    let custom_section_length = (custom_section_contents.len()
+        + custom_section_name.as_bytes().len()
+        + name_length_as_leb128.len()) as u32;
+
+    let mut custom_section_length_as_leb128 = vec![];
+    write_u32_leb128(&mut custom_section_length_as_leb128, custom_section_length);
+
+    out.extend(&custom_section_length_as_leb128);
+    out.extend(&name_length_as_leb128);
+    out.extend(custom_section_name.as_bytes());
+    out.extend(custom_section_contents);
+
+    out
+}
+
+#[test]
+fn it_works() {
+    use wasmer::wasm::{Global, Value};
+    use wasmer::{export, func, imports, CompiledModule, Func, Module, Table};
+    let wasm_no_custom_section = wabt::wat2wasm(TEST_WAT).unwrap();
+    let wasm = append_custom_section(
+        &wasm_no_custom_section,
+        "test_custom_section",
+        b"hello, world!",
+    );
+    let wasm = append_custom_section(&wasm, "test_custom_section", b"goodbye, world!");
+    // TODO: review error messages when `CompiledModule` is not in scope. My hypothesis is that they'll be
+    // misleading, if so we may want to do something about it.
+    let module = Module::new(wasm).unwrap();
+    let imports = module.imports();
+    assert_eq!(imports.len(), 2);
+    let num_exports = module.exports().count();
+    assert_eq!(num_exports, 6);
+    let (
+        mut test_table_seen,
+        mut test_global_seen,
+        mut test_ret_2_seen,
+        mut test_ret_4_seen,
+        mut set_test_global_seen,
+        mut update_outside_global_seen,
+    ) = (false, false, false, false, false, false);
+    for export in module.exports() {
```

Comment:
Shouldn't be better to just to have the exports to derive from `Eq`, `PartialEq` and then do a comparison on the expected exports via equality?

```rust
assert_eq!(module.exports(), [export::Export{kind:ExportKind::Table}, ...]);
```

Reply:
Hmm, I believe that would require them to either be orderable or hashable so we can compare them in an order-agnostic way but that wouldn't give us the property that we see every export exactly once. This seemed like a relatively simple way to completely test it in a way that's order agnostic (I didn't want the test to break if we change an implementation detail and the order they're returned in changes)

Reply:
I think exports should always return things in the proper order (as they are defined in the wasm).

Reply:
Sure, I'll update the docs on `ModuleInfo.exports` to say that it's supposed to be ordered and that we shouldn't use the APIs from ordered map that make it not insertion ordered.